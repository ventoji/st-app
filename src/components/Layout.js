import React from 'react';

const Layout = ({ className = 'container', children }) => (
  <div className={className}>{children}</div>
);

export default Layout;
