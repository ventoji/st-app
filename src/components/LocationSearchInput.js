/**
 * LocationSearchInput is a component from the third party library
 * https://www.npmjs.com/package/react-places-autocomplete
 * handleSelect and handleChange were customized for the app to handle
 * state redux.
 */
import React from 'react';
import PlacesAutocomplete, {
  geocodeByAddress,
  getLatLng,
} from 'react-places-autocomplete';

class LocationSearchInput extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      address: '',
      error: '',
    };
  }

  handleChange = (address) => {
    this.setState({ address });
    if (this.props.onSearchChange) this.props.onSearchChange({ address });
  };

  handleSelect = (address) => {
    geocodeByAddress(address)
      .then((results) => getLatLng(results[0]))
      .then((latLng) => {
        console.log('Success', latLng);
        if (this.props.onSelectChange)
          this.props.onSelectChange({ address, latLng });
      })
      .catch((error) => {
        this.setState({
          error: 'Connecting to Google Maps Api Failed',
        });
      });
  };

  render() {
    return (
      <PlacesAutocomplete
        value={this.state.address}
        onChange={this.handleChange}
        onSelect={this.handleSelect}
      >
        {({ getInputProps, suggestions, getSuggestionItemProps, loading }) => (
          <div>
            <input
              {...getInputProps({
                placeholder: 'Search Places ...',
                className: 'location-search-input',
              })}
            />
            <div className="autocomplete-dropdown-container">
              {loading && <div>Loading...</div>}
              {suggestions.map((suggestion) => {
                const className = suggestion.active
                  ? 'suggestion-item--active'
                  : 'suggestion-item';
                // inline style for demonstration purpose
                const style = suggestion.active
                  ? {
                      backgroundColor: '#fafafa',
                      cursor: 'pointer',
                      fontWeight: 'bold',
                    }
                  : { backgroundColor: '#ffffff', cursor: 'pointer' };
                return (
                  <div
                    {...getSuggestionItemProps(suggestion, {
                      className,
                      style,
                    })}
                  >
                    <span>{suggestion.description}</span>
                  </div>
                );
              })}
            </div>
          </div>
        )}
      </PlacesAutocomplete>
    );
  }
}

export default LocationSearchInput;
