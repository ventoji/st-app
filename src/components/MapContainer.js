import React, { Component } from 'react';
import { Map, InfoWindow, Marker, GoogleApiWrapper } from 'google-maps-react';
import shouldPureComponentUpdate from 'react-pure-render/function';
import { connect } from 'react-redux';
import { setLocationToSearch, setLocationsHistory } from '../actions/locations';
import { bindActionCreators } from 'redux';

export class MapContainer extends Component {
  constructor(props) {
    super(props);

    this.state = {
      selectedPlace: {},
      showingInfoWindow: false,
      activeMarker: {},
      mapStyles: {
        width: `${this.props.width}` || '100%',
        height: `${this.props.height}` || '100%',
      },
      stores: [], // markers asscoiated to every new search
      hideNav: false,
      zoom: this.props.zoom,
      center: this.props.center,
    };
  }

  shouldComponentUpdate = shouldPureComponentUpdate;

  componentDidMount() {
    window.addEventListener('resize', this.resize.bind(this));
    this.resize();
  }
  // Detect changes on the input search and update it properly
  componentDidUpdate(prevProps, prevState) {
    /*    if (prevProps.google !== this.props.google) {
      console.log('component has changed');
    } */

    if (prevProps.places !== this.props.places) {
      /*       console.log(
        'PLACES HAS CHAGED',
        this.props.places[this.props.places.length - 1],
      ); */

      this.setState(
        {
          stores: [
            ...this.state.stores,
            { ...this.props.places[this.props.places.length - 1] },
          ],
          center: {
            lat: this.props.places[this.props.places.length - 1].lat,
            lng: this.props.places[this.props.places.length - 1].lng,
          },
        },
        () => {
          // update marker after a new search is made
          this.displayMarkers();
          //console.log(this.state.stores, this.state.center);
        },
      );
    } // end check prev props

    if (this.state.hideNav) {
      this.setState({ zoom: 15 });
    } else {
      this.setState({ zoom: 5 });
    }
  }

  resize() {
    let currentHideNav = window.innerWidth <= 760;
    //  console.log(currentHideNav, window.innerWidth);
    if (currentHideNav !== this.state.hideNav) {
      this.setState({ hideNav: currentHideNav });
    }
  }

  onInfoWindowClose = () => {
    //   console.log('click Info');
    if (this.state.showingInfoWindow) {
      this.setState({
        showingInfoWindow: false,
        activeMarker: null,
      });
    }
  };
  onMarkerClick = (props, marker) => {
    this.setState({
      selectedPlace: props,
      activeMarker: marker,
      showingInfoWindow: true,
    });
  };

  displayMarkers = () =>
    this.state.stores.map((store, index) => (
      //  Show all markers, when an user makes a new search this is stored on
      // redux store
      <Marker
        key={index}
        id={index}
        onClick={this.onMarkerClick}
        position={{
          lat: store.lat,
          lng: store.lng,
        }}
        name={store.place}
      ></Marker>
    ));

  render() {
    return (
      <Map
        google={this.props.google}
        zoom={this.state.zoom}
        initialCenter={this.state.center}
      >
        {this.displayMarkers()}

        <InfoWindow
          marker={this.state.activeMarker}
          visible={this.state.showingInfoWindow}
          onClose={this.onInfoWindowClose}
        >
          <div>
            <h1>{this.state.selectedPlace.name}</h1>
          </div>
        </InfoWindow>
      </Map>
    );
  }
}

export const mapStateToProps = (state) => ({
  places: state.places,
});

export const mapDispatchToProps = (dispatch) =>
  bindActionCreators({ setLocationToSearch, setLocationsHistory }, dispatch);

export const GoogleApi = connect(
  mapStateToProps,
  mapDispatchToProps,
)(MapContainer);

export default GoogleApiWrapper({
  apiKey: process.env.APP_KEY_GOOGLE_MAPS,
})(GoogleApi);
