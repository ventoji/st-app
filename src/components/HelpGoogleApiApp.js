import React from 'react';
import Layout from '../components/Layout';

const HelpExpensePage = () => (
  <Layout>
    <p> quicly explanation how to use it</p>;
  </Layout>
);

export default HelpExpensePage;
