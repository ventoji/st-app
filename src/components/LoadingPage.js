import React from 'react';
import Layout from '../components/Layout';

const LoadingPage = () => (
  <Layout className="loader">
    <img className="loader__image" src="/images/loader.gif" alt="loader" />
  </Layout>
);

export default LoadingPage;
