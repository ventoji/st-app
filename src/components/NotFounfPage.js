import React from 'react';
import { Link } from 'react-router-dom';
import Layout from '../components/Layout';

const NotFoundPage = () => (
  <Layout>
    404 - <Link to="/"> Go Home </Link>
  </Layout>
);

export default NotFoundPage;
