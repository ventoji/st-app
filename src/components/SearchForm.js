import React, { useState } from 'react';
import LocationSearchInput from '../components/LocationSearchInput';
//import TestingComponent from '../components/TestingComponent';

// Handle input search text by the user
export const SearchForm = ({ setLocationsHistory, setLocationToSearch }) => {
  const [querySearchText, setQuerySearchText] = useState('');

  const onSelectChange = (e) => {
    if (querySearchText) {
      setLocationsHistory(querySearchText, { place: e.address, ...e.latLng });
    }
  };

  const onSearchChange = (e) => {
    setLocationToSearch(e.address);
    setQuerySearchText(e.address);
  };

  return (
    <LocationSearchInput
      onSearchChange={onSearchChange}
      onSelectChange={onSelectChange}
    />
  );
};
