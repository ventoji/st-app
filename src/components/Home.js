/**
 *  Render the main view for the App
 *  SearchLocation a connected component to display the Search form
 *  MapContainer: Display the Google Map.
 */
import React from 'react';
import Layout from '../components/Layout';
import SearchLocation from '../container/SearchLocation';
import { propsGoogleMap } from '../utils/index';
import MapContainer from '../components/MapContainer';

const Home = () => (
  <Layout className="main-container">
    <div className="search">
      <div className="search__form">
        {' '}
        <SearchLocation />
      </div>
    </div>
    <div className="map-container">
      <MapContainer {...propsGoogleMap} />
    </div>
  </Layout>
);

export default Home;
