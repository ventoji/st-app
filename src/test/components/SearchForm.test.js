import React from 'react';
import { shallow } from 'enzyme';
import { SearchForm } from '../../components/SearchForm';

let setLocationsHistory, setLocationToSearch, wrapper;
beforeEach(() => {
  setLocationsHistory = jest.fn();
  setLocationToSearch = jest.fn();
  wrapper = shallow(
    <SearchForm
      setLocationsHistory={setLocationsHistory}
      setLocationToSearch={setLocationToSearch}
    />,
  );
});

test('should render SearchForm correctly', () => {
  expect(wrapper).toMatchSnapshot();
});

test.skip('should handle onSelectChange ', () => {
  wrapper.find('LocationSearchInput').prop('onSelectChange')('Madrid');
  expect(setLocationsHistory).toHaveBeenCalled();
});

test('should handle onSearchChange ', () => {
  wrapper.find('LocationSearchInput').prop('onSearchChange')('Ma');
  expect(setLocationToSearch).toHaveBeenCalled();
  expect(setLocationToSearch).toHaveBeenCalledTimes(1);
});
