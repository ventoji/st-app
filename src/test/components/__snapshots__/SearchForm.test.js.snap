// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render SearchForm correctly 1`] = `
<LocationSearchInput
  onSearchChange={[Function]}
  onSelectChange={[Function]}
/>
`;
